### DISCUSSION TITLE
能不能替我释疑一下这块的注释

### DISCUSSION THREAD

**Discussion by dev4u** [shadowsocks/shadowsocks-rust#762] [General]
https://github.com/shadowsocks/shadowsocks-rust/blob/0e0edda77b4bb43ce49df0e0f487d10bde168d29/crates/shadowsocks-service/src/acl/mod.rs#L479-L489

这块代码按照注释的描述，`Mode::BlackList`表示当前模式下，默认以代理方式处理请求。`Mode::WhiteList`则相反，默认以直连处理请求。

https://github.com/shadowsocks/shadowsocks-rust/blob/0e0edda77b4bb43ce49df0e0f487d10bde168d29/crates/shadowsocks-service/src/acl/mod.rs#L30-L35

但是在`Mode`的定义里，代码描述的注释却是与上面的逻辑是相反的：`Mode::BlackList`默认直连，`Mode::WhiteList`默认代理。

这块我搞不明白要怎么理解，能不能给我解释一下呢？

**Comment by zonyitoo** [comment]
If mode is BlackList, then domains that are configured should be bypassed. So, the others are “defaults” to be in proxied list.

**Reply by dev4u** [reply]
在sslocal下，按照你的解释，上下两段代码对blacklist的逻辑是不统一。
你说的blacklist对应的acl是proxy_all模式，

但是在代码上的注释却是:`BlackList mode, rejects or bypasses all requests by default`，这个应该是bypass_all模式才对。

**Reply by zonyitoo** [reply]
https://github.com/shadowsocks/shadowsocks-rust/blob/46934f5675b0554cad52c8884787519d7a4bee94/crates/shadowsocks-service/src/acl/mod.rs#L288-L307

`proxy_all` is `WhiteList`.

**Reply by dev4u** [reply]
感谢解答，我会继续走读代码。

**Comment by dev4u** [comment]
请教一下，我这边想实现sslocal<->ssserver之间的pingpong交互，有了这个交互，两边的应用可以实现数据交换。
这个逻辑我可以从哪里入手呢？

**Reply by zonyitoo** [reply]
Add a protocol header in the client and server implementation in the `shadowsocks` crate.

**Reply by dev4u** [reply]
好的，我去看看。

**Comment by dev4u** [comment]
能不能提供一个工具，可以让我测试手机在不同基站、或者不同网络内漫游切换的时候，有没发生访问不到网络的情况发生？
